File: app/src/main/java/com/example/tree/Co2CalActivity.java
package com.example.tree;


import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.view.GravityCompat;
import androidx.drawerlayout.widget.DrawerLayout;


import android.content.Context;
import android.content.Intent;
import android.location.Address;
import android.location.Geocoder;
import android.location.Location;
import android.os.Bundle;
import android.util.Log;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import com.example.tree.Main.DataModels_Main;
import com.example.tree.Main.MainData_response;
import com.google.android.gms.maps.model.LatLng;
import com.example.tree.Main.MainTreeActivity;
import com.example.tree.Main.RetrofitAPI;

import org.mozilla.javascript.tools.jsc.Main;

import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class Co2CalActivity extends AppCompatActivity {
    private DrawerLayout mDrawerLayout;
    private Context context = this;

    private final String BASE_URL = "http://180.230.121.23/";
    private RetrofitAPI mMyAPI;

    //로그캣 사용 설정
    private static final String TAG = "Co2CalActivity";


    Button btnKor3Loc, btnKor2Loc, d_btn,give_water_btn;
    EditText editText, editText2;
    TextView text1 = null;
    double distance;
    int check;
    String meter;
    Location location1 = new Location("point 1");
    Location location2 = new Location("point 2");
    LatLng curPoint1;
    LatLng curPoint2;
    DrawerLayout drawerLayout;

    //Co2배출량
    float emission=0;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_co2_cal);

        drawerLayout=findViewById(R.id.drawer_layout);
        //권한 설정
        //checkDangerousPermissions();
        initMyAPI(BASE_URL);
        //객체 초기화
        editText = findViewById(R.id.editText);
        btnKor2Loc = findViewById(R.id.button2);
        editText2 = findViewById(R.id.editText2);
        btnKor3Loc = findViewById(R.id.button3);
        text1 = (TextView) findViewById(R.id.distance);
        d_btn = findViewById(R.id.d_btn);
        give_water_btn=findViewById(R.id.give_water_btn);


        btnKor2Loc.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (editText.getText().toString().length() > 0) {
                    location1 = getLocationFromAddress(getApplicationContext(), editText.getText().toString());
                    curPoint1 = new LatLng(location1.getLatitude(), location1.getLongitude());
                    location1.setLatitude(curPoint1.latitude);
                    location1.setLongitude(curPoint1.longitude);
                }
            }
        });

        btnKor3Loc.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (editText2.getText().toString().length() > 0) {
                    location2 = getLocationFromAddress(getApplicationContext(), editText2.getText().toString());
                    curPoint2 = new LatLng(location2.getLatitude(), location2.getLongitude());
                    location2.setLatitude(curPoint2.latitude);
                    location2.setLongitude(curPoint2.longitude);
                }
            }
        });

        d_btn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                distance = location1.distanceTo(location2);
                distance = (float) Math.round(distance * 0.001);
                meter = Double.toString(distance);
                Log.d(TAG, "DistanceBetweenPointAandB: " + distance);
                text1.setText("이동 거리(km) : " + meter + "km");
            }
        });
        //체크박스
        CheckBox checkBox1 = (CheckBox) findViewById(R.id.check1) ;
        CheckBox checkBox2 = (CheckBox) findViewById(R.id.check2) ;
        CheckBox checkBox3 = (CheckBox) findViewById(R.id.check3) ;
        checkBox1.setOnClickListener(new CheckBox.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (((CheckBox)v).isChecked()) {
                    // TODO : CheckBox is checked.
                    check=0;
                }
            }
        });
        checkBox2.setOnClickListener(new CheckBox.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (((CheckBox)v).isChecked()) {
                    // TODO : CheckBox is checked.
                    check=1;
                }
            }
        });
        checkBox3.setOnClickListener(new CheckBox.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (((CheckBox)v).isChecked()) {
                    // TODO : CheckBox is checked.
                    check=2;
                }
            }
        });

        give_water_btn.setOnClickListener(new View.OnClickListener(){
            public void onClick(View v) {
                    Log.d(TAG, "POST");
                    //RetrofitAPI postAPI = retrofit.create(RetrofitAPI.class);
                    DataModels_Main data=new DataModels_Main(distance,check);
               /* RetrofitAPI result=mMyAPI;
                result.responsePost(distance, check).enqueue(new Callback<DataModels_Main>() {
                    @Override
                    public void onResponse(@NonNull Call<DataModels_Main> call, @NonNull Response<DataModels_Main> response) {
                        if(response.isSuccessful()) {  // 조회성공
                            DataModels_Main res = response.body();
                            Log.d("총 누적 배출량 ", String.valueOf(res.getTotalReduction()));
                            Log.d("트리 레벨", String.valueOf(res.getTreeLevel()));
                            Log.d("트리 개수", String.valueOf(res.getTreeCount()));
                            Log.d("레벨 별 배출량", String.valueOf(res.getLevelReduction()));
                        }
                    }
                    @Override
                    public void onFailure(Call<DataModels_Main> call, Throwable t) {

                        t.printStackTrace();
                    }
                });*/

                Call<MainData_response> postCall = mMyAPI.createPost(data);
                postCall.enqueue(new Callback<MainData_response>() {
                    @Override
                    public void onResponse(Call<MainData_response> call, Response<MainData_response> response) {
                        Log.d("retrofit", "Data fetch success");
                        if (response.isSuccessful()) {
                            MainData_response result=response.body();
                            Log.d(TAG, "등록 완료");
                            Intent intent =new Intent(getBaseContext(), MainTreeActivity.class);
                            intent.putExtra("totalReduction", result.getTotalReduction());
                            startActivity(intent);
                        } else {
                            Log.d(TAG, "Status Code : " + response.code());
                            Log.d(TAG, response.errorBody().toString());
                            Log.d(TAG, call.request().body().toString());
                        }
                    }

                    @Override
                    public void onFailure(Call<MainData_response> call, Throwable t) {
                        Log.d(TAG, "Fail msg : " + t.getMessage());
                        Toast.makeText(getApplicationContext(), "FAIL", Toast.LENGTH_SHORT).show();
                    }
                });
            }

        });
    }

    private void initMyAPI(String base_url) {
        Log.d(TAG,"initMyAPI : " + base_url);
        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(base_url)
                .addConverterFactory(GsonConverterFactory.create())
                .build();

        mMyAPI = retrofit.create(RetrofitAPI.class);

    }




    public void ClickMenu(View view){
        MainTreeActivity.openDrawer(drawerLayout);
    }
    public void ClickLogo(View view){
        MainTreeActivity.closeDrawer(drawerLayout);
    }

    public void ClickHome(View view){
        MainTreeActivity.redirectActivity(this, MainTreeActivity.class);
    }

    public void ClickDashboard(View view){
        recreate();
    }

    public void ClickAboutUs(View view){
        MainTreeActivity.redirectActivity(this,Co2CalActivity.class);
    }
    public void Neighborhood(View view){
        MainTreeActivity.redirectActivity(this,NeighborhoodActivity.class);
    }
    public void Setting(View view){
        MainTreeActivity.redirectActivity(this,SettingActivity.class);
    }

    protected void onPause(){
        super.onPause();
        MainTreeActivity.closeDrawer(drawerLayout);
    }


    private Location getLocationFromAddress(Context context, String address) {
        Geocoder geocoder = new Geocoder(context);
        List<Address> addresses;
        Location resLocation = new Location("");
        try {
            addresses = geocoder.getFromLocationName(address, 5);
            if ((addresses == null) || (addresses.size() == 0)) {
                return null;
            }
            Address addressLoc = addresses.get(0);

            resLocation.setLatitude(addressLoc.getLatitude());
            resLocation.setLongitude(addressLoc.getLongitude());

        } catch (Exception e) {
            e.printStackTrace();
        }
        return resLocation;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home: { // 왼쪽 상단 버튼 눌렀을 때
                mDrawerLayout.openDrawer(GravityCompat.START);
                return true;
            }
        }
        return super.onOptionsItemSelected(item);
    }

//    private void showCurrentLocation(Location location) {
//        LatLng curPoint = new LatLng(location.getLatitude(), location.getLongitude());
//        String msg = "Latitutde : " + curPoint.latitude
//                + "\nLongitude : " + curPoint.longitude;
//        //Toast.makeText(this, msg, Toast.LENGTH_SHORT).show();
//    }

}

File: app/src/main/java/com/example/tree/Co2CalActivity.kt
package com.example.tree

import android.content.Context
import android.content.Intent
import android.location.Address
import android.location.Geocoder
import android.location.Location
import android.os.Bundle
import android.util.Log
import android.view.MenuItem
import android.view.View
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.GravityCompat
import androidx.drawerlayout.widget.DrawerLayout
import com.example.tree.Main.*
import com.example.tree.Main.MainTreeActivity.Companion.closeDrawer
import com.example.tree.Main.MainTreeActivity.Companion.openDrawer
import com.example.tree.Main.MainTreeActivity.Companion.redirectActivity
import com.example.tree.neighbor.DeleteModel
import com.google.android.gms.maps.model.LatLng
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

class Co2CalActivity : AppCompatActivity() {
    private val mDrawerLayout: DrawerLayout? = null
    private val context: Context = this

    var distance : Double=0.0
    var check : Int = 0
    var meter: String? = null
    var location1: Location? = Location("point 1")
    var location2: Location? = Location("point 2")
    var curPoint1: LatLng? = null
    var curPoint2: LatLng? = null
    var drawerLayout: DrawerLayout? = null


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_co2_cal)
        drawerLayout = findViewById(R.id.drawer_layout)
        //객체 초기화
        var editText : EditText = findViewById(R.id.editText)
        var btnKor2Loc : Button= findViewById(R.id.button2)
        var editText2 : EditText= findViewById(R.id.editText2)
        var btnKor3Loc : Button= findViewById(R.id.button3)
        var text1 :TextView= findViewById<View>(R.id.distance) as TextView
        var d_btn :Button= findViewById(R.id.d_btn)
        var give_water_btn :Button= findViewById(R.id.give_water_btn)


        btnKor2Loc.setOnClickListener(View.OnClickListener {
            if (editText.getText().toString().length > 0) {
                location1 =
                    getLocationFromAddress(applicationContext, editText.getText().toString())
                curPoint1 = LatLng(location1!!.latitude, location1!!.longitude)
                location1!!.latitude = curPoint1!!.latitude
                location1!!.longitude = curPoint1!!.longitude
            }
        })
        btnKor3Loc.setOnClickListener(View.OnClickListener {
            if (editText2.getText().toString().length > 0) {
                location2 =
                    getLocationFromAddress(applicationContext, editText2.getText().toString())
                curPoint2 = LatLng(location2!!.latitude, location2!!.longitude)
                location2!!.latitude = curPoint2!!.latitude
                location2!!.longitude = curPoint2!!.longitude
            }
        })

        d_btn.setOnClickListener(View.OnClickListener {
            distance = location1!!.distanceTo(location2).toDouble()
            distance = Math.round(distance * 0.001).toDouble()
            meter = java.lang.Double.toString(distance)
            Log.d(TAG, "DistanceBetweenPointAandB: $distance")
            text1!!.text = "이동 거리(km) : " + meter + "km"
        })
        //체크박스
        val checkBox1 = findViewById<View>(R.id.check1) as CheckBox
        val checkBox2 = findViewById<View>(R.id.check2) as CheckBox
        val checkBox3 = findViewById<View>(R.id.check3) as CheckBox
        checkBox1.setOnClickListener { v ->
            if ((v as CheckBox).isChecked) {
                // TODO : CheckBox is checked.
                check = 0
            }
        }
        checkBox2.setOnClickListener { v ->
            if ((v as CheckBox).isChecked) {
                // TODO : CheckBox is checked.
                check = 1
            }
        }
        checkBox3.setOnClickListener { v ->
            if ((v as CheckBox).isChecked) {
                // TODO : CheckBox is checked.
                check = 2
            }
        }

        give_water_btn.setOnClickListener(View.OnClickListener {
            Log.d(TAG, "POST")
            //RetrofitAPI postAPI = retrofit.create(RetrofitAPI.class);
            val data = Data_Request(distance, check)
            /* RetrofitAPI result=mMyAPI;
                    result.responsePost(distance, check).enqueue(new Callback<DataModels_Main>() {
                        @Override
                        public void onResponse(@NonNull Call<DataModels_Main> call, @NonNull Response<DataModels_Main> response) {
                            if(response.isSuccessful()) {  // 조회성공
                                DataModels_Main res = response.body();
                                Log.d("총 누적 배출량 ", String.valueOf(res.getTotalReduction()));
                                Log.d("트리 레벨", String.valueOf(res.getTreeLevel()));
                                Log.d("트리 개수", String.valueOf(res.getTreeCount()));
                                Log.d("레벨 별 배출량", String.valueOf(res.getLevelReduction()));
                            }
                        }
                        @Override
                        public void onFailure(Call<DataModels_Main> call, Throwable t) {
    
                            t.printStackTrace();
                        }
                    });*/
            (application as MasterApplication).service_tree.createPost(data)
                .enqueue(object : Callback<Data_Response> {
                override fun onResponse(
                    call: Call<Data_Response>,
                    response: Response<Data_Response>
                ) {
                    Log.d("retrofit", "Data fetch success")
                    if (response.isSuccessful) {
                        val result = response.body()
                        Log.d(TAG, "등록 완료")
                        val intent = Intent(baseContext, MainTreeActivity::class.java)
                        intent.putExtra("totalReduction", result!!.totalReduction)
                        startActivity(intent)
                    } else {
                        Log.d(TAG, "Status Code : " + response.code())
                        Log.d(TAG, response.errorBody().toString())
                        Log.d(TAG, call.request().body().toString())
                    }
                }

                override fun onFailure(call: Call<Data_Response>, t: Throwable) {
                    Log.d(TAG, "Fail msg : " + t.message)
                    Toast.makeText(applicationContext, "FAIL", Toast.LENGTH_SHORT).show()
                }
            })
        })
    }

    fun ClickMenu(view: View?) {
        openDrawer(drawerLayout)
    }

    fun ClickLogo(view: View?) {
        closeDrawer(drawerLayout)
    }

    fun ClickHome(view: View?) {
        redirectActivity(this, MainTreeActivity::class.java)
    }

    fun ClickDashboard(view: View?) {
        recreate()
    }

    fun ClickAboutUs(view: View?) {
        redirectActivity(this, allRankingActivity::class.java)
    }

    fun Neighborhood(view: View?) {
        redirectActivity(this, NeighborhoodActivity::class.java)
    }

    fun Setting(view: View?) {
        redirectActivity(this, SettingActivity::class.java)
    }

    override fun onPause() {
        super.onPause()
        closeDrawer(drawerLayout)
    }

    private fun getLocationFromAddress(context: Context, address: String): Location? {
        val geocoder = Geocoder(context)
        val addresses: List<Address>?
        val resLocation = Location("")
        try {
            addresses = geocoder.getFromLocationName(address, 5)
            if (addresses == null || addresses.size == 0) {
                return null
            }
            val addressLoc = addresses[0]
            resLocation.latitude = addressLoc.latitude
            resLocation.longitude = addressLoc.longitude
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return resLocation
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                // 왼쪽 상단 버튼 눌렀을 때
                mDrawerLayout!!.openDrawer(GravityCompat.START)
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    companion object {
        //로그캣 사용 설정
        private const val TAG = "Co2CalActivity"
    }
}